fabric-transaction-api-v1/src/main/java/net/fabricmc/fabric/impl/transaction/TransactionImpl.java

```diff
@@ -0,0 +1,193 @@
+	}
+
+	private final ContextImpl context = new ContextImpl();
+	private boolean isOpen = true;
+	private boolean isCommited = false;
+	private final IdentityHashMap<TransactionDelegate, Consumer<TransactionContext>> participants = new IdentityHashMap<>();
+	private final IdentityHashMap<TransactionDelegate, Object> stateStorage = new IdentityHashMap<>();
+	private TransactionDelegate contextDelegate;
+
+	private TransactionImpl() {
+	}
+
+	@Override
+	public void close() {
+		if (isOpen) {
+			rollback();
+		}
+	}
+
+	private void clear() {
+		participants.clear();
+		stateStorage.clear();
+		contextDelegate = null;
+		isOpen = false;
+		isCommited = false;
+	}
+
+	private void validate() {
+		if (!isOpen) {
+			throw new IllegalStateException("Encountered transaction operation for closed transaction.");
+		}
+
+		if (STACK.get(stackPointer) != this) {
+			throw new IndexOutOfBoundsException("Transaction operations must apply to most recent open transaction.");
+		}
+
+		if (!innerLock.isHeldByCurrentThread()) {
+			throw new ConcurrentModificationException("Attempt to modify transaction status from foreign thread");
+		}
+	}
+
+	@Override
+	public void rollback() {
+		close(false);
+	}
+
+	@Override
+	public void commit() {
+		close(true);
+	}
+
+	private void close(boolean isCommited) {
+		validate();
+		this.isCommited = isCommited;
+
+		participants.forEach((c, r) -> {
+			contextDelegate = c;
+			r.accept(context);
+		});
+
+		clear();
+
+		final boolean root = --stackPointer == -1;
+
+		innerLock.unlock();
+
+		// non-server threads have an additional lock we must release.
+		if (Thread.currentThread() != serverThread) {
+			outerLock.unlock();
+
+			if (root) {
+				// Give other threads (like the server thread) that want to do transactions a
+				// chance to jump in.
+				Thread.yield();
+			}
+		}
+	}
+
+	@Override
+	public <T extends TransactionParticipant> T enlistSelf(T container) {
+		validate();
+		final TransactionDelegate d = container.getTransactionDelegate();
+
+		if (!participants.containsKey(d)) {
+			contextDelegate = d;
+			participants.put(d, defaultRollback(d.prepareCompletionHandler(context)));
+			contextDelegate = null;
+		}
+
+		return container;
+	}
+
+	private Consumer<TransactionContext> defaultRollback(Consumer<TransactionContext> consumer) {
+		return consumer;
+	}
+
+	///// STATIC MEMBERS FOLLOW /////
+
+	public static void setServerThread(MinecraftServer server) {
+		serverThread = server == null ? null : Thread.currentThread();
+	}
+
+	private static Thread serverThread;
+
+	/** Held by thread of active transaction, regardless of thread. */
+	static final ReentrantLock innerLock = new ReentrantLock();
+
+	/**
+	 * Must be held if thread is not the server thread - ensures non-server threads
+	 * contend with each other before contending with server thread itself.
+	 */
+	static final ReentrantLock outerLock = new ReentrantLock();
+
+	private static final ArrayList<TransactionImpl> STACK = new ArrayList<>();
```

Comment:
can't you use a stack data structure?

Reply:
I want to retain and reuse instances in this case because otherwise re-allocation could be frequent.  Could also achieve that with a stack if I stash popped references in a different structure for reuse, but this seems like less machinery overall.